core/verify_manager.go

```diff
@@ -0,0 +1,134 @@
+package core
+
+import (
+	"fmt"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const verifiedCacheSize = 256
+
+type VerifyManager struct {
+	bc                   *BlockChain
+	tasks                map[common.Hash]*VerifyTask
+	peers                VerifyPeers
+	verifiedCache        *lru.Cache
+	allowUntrustedVerify bool
+	verifyCh             chan common.Hash
+	exitCh               chan struct{}
+}
+
+func NewVerifyManager(blockchain *BlockChain) *VerifyManager {
+	verifiedCache, _ := lru.New(verifiedCacheSize)
+	vm := &VerifyManager{
+		bc:            blockchain,
+		tasks:         make(map[common.Hash]*VerifyTask),
+		verifiedCache: verifiedCache,
+		verifyCh:      make(chan common.Hash),
+		exitCh:        make(chan struct{}),
+	}
+	return vm
+}
+
+func (vm *VerifyManager) Start() {
+	//read disk store to initial verified cache
+	//load unverified blocks in a normalized chain and start a batch of verify task
+	header := vm.bc.CurrentHeader()
+	vm.NewBlockVerifyTask(header)
+	go vm.mainLoop()
+}
+
+func (vm *VerifyManager) Stop() {
+	defer close(vm.exitCh)
+	vm.exitCh <- struct{}{}
+}
+
+func (vm *VerifyManager) mainLoop() {
+	prune := time.NewTicker(time.Second)
+	defer prune.Stop()
+	for {
+		select {
+		case hash := <-vm.verifyCh:
+			vm.cacheBlockVerified(hash)
+			rawdb.WriteTrustBlockHash(vm.bc.db, hash)
+			delete(vm.tasks, hash)
+		case <-prune.C:
+			for hash, task := range vm.tasks {
+				if vm.bc.CurrentHeader().Number.Uint64()-task.blockHeader.Number.Uint64() > 15 {
+					task.terminalCh <- struct{}{}
+					delete(vm.tasks, hash)
+				}
+			}
+		case <-vm.exitCh:
+			return
+		}
+	}
+}
+
+func (vm *VerifyManager) NewBlockVerifyTask(header *types.Header) {
+	for i := 0; i < 11; i++ {
+		hash := header.Hash()
+		diffLayer := vm.bc.GetTrustedDiffLayer(hash)
+		//if this block has no diff, there is no need to verify it.
+		if diffLayer == nil {
+			continue
+		}
+		//if verified cache record that this block has been verified, skip.
+		if _, ok := vm.verifiedCache.Get(hash); ok {
+			continue
+		}
+		//if verified storage record that this block has been verified, skip.
+		if rawdb.ReadTrustBlockHash(vm.bc.db, hash) {
+			vm.cacheBlockVerified(hash)
+			continue
+		}
+		//if there already has a verify task for this block, skip.
+		if _, ok := vm.tasks[hash]; ok {
+			continue
+		}
+		diffHash, err := GetTrustedDiffHash(diffLayer)
+		if err != nil {
+			log.Error("failed to get diff hash", "block", hash, "number", header.Number, "error:", err)
+		}
+		verifyTask := NewVerifyTask(diffHash, header, vm.peers, vm.bc.db, vm.verifyCh, vm.allowUntrustedVerify)
+		vm.tasks[hash] = verifyTask
+		header = vm.bc.GetHeaderByHash(header.ParentHash)
```

Comment:
If the node runs from genesis block, the header would be nil, we should handle this case.

Reply:
done

Reply:
I did not see it been fixed?